From f35800070167593240ee15a8e07c0e248b6eb5f5 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Mon, 07 Oct 2019 05:45:45 +0300
Subject: [PATCH] [SW-0] gaudi: add BTL pcie enable logic

This logic copied from BTL PCIe section (#4)
In BTL it is called when DIP switch for PCIe enable is set
Now we wish to do this logic also here as a measure of security
It means that in new PCB's version by default PCIe won't be enabled
So without proper FW device won't be accessible through PCIe
Hence device is more secured

Note: we assume new PCB will include Metal Fix hence no need for calling
init_pll_pci logic

Change-Id: I85c2ec63528e781e3e397af521dd67ae7eb03b89
---

diff --git a/gaudi/common.c b/gaudi/common.c
index fd6b559..017cf91 100644
--- a/gaudi/common.c
+++ b/gaudi/common.c
@@ -8,6 +8,12 @@
 #include <psoc_global_conf_regs.h>
 #include <pcie_aux_regs.h>
 #include <pcie_core_regs.h>
+#include <pcie_wrap_regs.h>
+#include <pcie_phy_regs.h>
+#include <pcie_dbi_regs.h>
+
+#include <hl_boot_if.h>
+#include <gaudi/gaudi_reg_map.h>
 
 #include <preboot.h>
 #include <pll.h>
@@ -15,12 +21,110 @@
 
 #include "common.h"
 
+#define FW_WORD_NUM		0x2000	/* FW_SIZE / 2 */
+static void preboot_init_pcie_phy_fw_update(void)
+{
+	/* Configure FW update size */
+	regmap_writel(FW_WORD_NUM, mmPCIE_WRAP_PHY_FW_FSM_SIZE);
+
+	/*
+	 * PHY FW FSM EN and Direct FW write mechanism enable
+	 * FW FSM assumes that the data is present in the SRAM
+	 * The FSM will wait for SRAM init done before updating
+	 * the SRAM contents
+	 */
+	regmap_writel(9, mmPCIE_WRAP_PHY_FW_FSM);
+
+	/* Take the PHY out of reset */
+	regmap_writel(1, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
+}
+
+static void preboot_init_pcie_phy(void)
+{
+	/* Reset PCIe PHY */
+	regmap_writel(0, mmPSOC_GLOBAL_CONF_PCIE_PHY_RST_N);
+
+	/*
+	 * Set PCIe PHY CR Para I/F to CFG.
+	 * Select CR PARA I/F instead of JTAG
+	 */
+	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_0);
+	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_1);
+	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_2);
+	regmap_writel(1, mmPCIE_PHY_CR_PARA_SEL_3);
+
+	/* Select indirect access to CR PARA I/F instead of PRBI */
+	regmap_writel(1, mmPCIE_PHY_CR_PARA_CFG);
+
+	/*
+	 * Use PHY configuration from SoC online
+	 * instead of default values
+	 */
+	regmap_writel(1, mmPCIE_PHY_PHY_EXT_CTRL_SEL);
+	preboot_init_pcie_phy_fw_update();
+}
+
+static void preboot_init_pcie(void)
+{
+	preboot_init_pcie_phy();
+
+	/* Update Vendor ID */
+	regmap_writel(0x10001DA3, mmPCIE_DBI_DEVICE_ID_VENDOR_ID_REG);
+
+	/* Update subsystem Vendor ID */
+	regmap_writel(0x10001DA3,
+		      mmPCIE_DBI_SUBSYSTEM_ID_SUBSYSTEM_VENDOR_ID_REG);
+
+	/* Disable the EQ REDO */
+	regmap_writel(0x00402801, mmPCIE_DBI_GEN3_RELATED_OFF);
+
+	/* delay before executing Synopsys W/A > 300us */
+	hl_pll_udelay(300);
+
+	/* Execute the Synopsys W/A */
+	regmap_writel(0xF, mmPCIE_PHY_CR_PARA_CTRL);
+	regmap_writel(0x0004A07D, mmPCIE_PHY_CR_PARA_CMD);
+	hl_pll_udelay(20);
+
+	/* Enable and Mask LTSSM (erratum 102) */
+	regmap_writel(3, mmPCIE_AUX_LTSSM_EN);
+}
+
+/*
+ * This logic copied from BTL PCIe section (#4)
+ * In BTL it is called when DIP switch for PCIe enable is set
+ * Now we wish to do this logic also here as a measure of security
+ * It means that in new PCB's version by default PCIe won't be enabled
+ * So without proper FW device won't be accessible through PCIe
+ * Hence device is more secured
+ */
+static void preboot_pcie_enable(void)
+{
+	u32 boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS);
+	u32 reset_cause;
+
+	/* Check if already done by BTL */
+	if (boot_strap & 0x2)
+		return;
+
+	/* During hard reset triggered by LKD keep PCIe intact */
+	reset_cause = regmap_readl(mmPREBOOT_PCIE_EN);
+	regmap_writel(0, mmPREBOOT_PCIE_EN);
+	if (reset_cause == LKD_HARD_RESET_MAGIC)
+		return;
+
+	preboot_init_pcie();
+}
+
 void preboot_pcie_wait_for_link(void)
 {
 	/* Wait for PRSTn */
 	while (!regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN))
 		;
 
+	/* Only after PRSTn is de-asserted */
+	preboot_pcie_enable();
+
 	while (1) {
 		unsigned int prstn, ltssm, smlh, rdlh;
 
